geth/peers/topicpool.go

```diff
@@ -0,0 +1,234 @@
+	period     chan time.Duration
+
+	cache *Cache
+}
+
+// SearchRunning returns true if search is running
+func (t *TopicPool) SearchRunning() bool {
+	return atomic.LoadInt32(&t.running) == 1
+}
+
+// MaxReached returns true if we connected with max number of peers.
+func (t *TopicPool) MaxReached() bool {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.connected == t.limits[1]
+}
+
+// ConfirmAdded called when peer was added by p2p Server
+func (t *TopicPool) ConfirmAdded(server *p2p.Server, nodeID discover.NodeID) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// inbound connection
+	peer, exist := t.peers[discv5.NodeID(nodeID)]
+	if !exist {
+		return
+	}
+	// established connection means that the node is a viable candidate for a connection and can be cached
+	if t.cache != nil {
+		if err := t.cache.AddPeer(peer.node, t.topic); err != nil {
+			log.Error("failed to persist a peer", "error", err)
+		}
+	}
+	// when max limit is reached drop every peer after
+	if t.connected == t.limits[1] {
+		log.Debug("max limit is reached drop the peer", "ID", nodeID, "topic", t.topic)
+		t.removePeer(server, peer)
+		return
+	}
+	// don't count same peer twice
+	if !peer.connected {
+		log.Debug("marking as connected", "ID", nodeID)
+		peer.connected = true
+		t.connected++
+	}
+	if t.SearchRunning() && t.connected == t.limits[0] {
+		t.period <- t.slowSync
+	}
+	return
+}
+
+// ConfirmDropped called when server receives drop event
+func (t *TopicPool) ConfirmDropped(server *p2p.Server, nodeID discover.NodeID, reason string) (new bool) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// either inbound or connected from another topic
+	peer, exist := t.peers[discv5.NodeID(nodeID)]
+	if !exist {
+		return false
+	}
+	log.Debug("disconnect reason", "peer", nodeID, "reason", reason)
+	// if requested - we don't need to remove peer from cache and look for a replacement
+	if reason == p2p.DiscRequested.Error() {
+		return false
+	}
+	if t.SearchRunning() && t.connected == t.limits[0] {
+		t.period <- t.fastSync
+	}
+	t.connected--
+	t.removePeer(server, peer)
+	delete(t.peers, discv5.NodeID(nodeID))
+	if t.cache != nil {
+		if err := t.cache.RemovePeer(discv5.NodeID(nodeID), t.topic); err != nil {
+			log.Error("failed to remove peer from cache", "error", err)
+		}
+	}
+	// TODO use a heap queue and always get a peer that was discovered recently
+	for _, peer := range t.peers {
+		if !peer.connected && mclock.Now() < peer.discoveredTime+mclock.AbsTime(expirationPeriod) {
+			t.addPeer(server, peer)
+			return true
+		}
+	}
+	return false
+}
+
+// StartSearch creates discv5 queries and runs a loop to consume found peers.
+func (t *TopicPool) StartSearch(server *p2p.Server) error {
+	if atomic.LoadInt32(&t.running) == 1 {
+		return nil
+	}
+	if server.DiscV5 == nil {
+		return ErrDiscv5NotRunning
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	atomic.StoreInt32(&t.running, 1)
+	t.quit = make(chan struct{})
+	t.period = make(chan time.Duration, 2)
```

Comment:
We need to provide comments why these channels are buffered and why these buffer sizes were chosen.